src/core/net/dhcp6_server.cpp

```diff
@@ -0,0 +1,544 @@
+
+    SuccessOrExit(SendReply(aDst, aType, aTransactionId, clientIdentifier, iana, iaAddress));
+
+exit:
+    {}
+}
+
+ThreadError Dhcp6Server::ProcessClientIdentifier(Message &aMessage, uint16_t aOffset, ClientIdentifier &aClient)
+{
+    ThreadError error = kThreadError_None;
+    aMessage.Read(aOffset, sizeof(aClient), &aClient);
+
+    VerifyOrExit(((aMessage.Read(aOffset, sizeof(aClient), &aClient) == sizeof(aClient)) &&
+                  (aClient.GetLength() == (sizeof(aClient) - sizeof(Dhcp6Option))) &&
+                  (aClient.GetDuidType() == kDuidLL) &&
+                  (HostSwap16(aClient.GetDuidLL().mDuidHardwareType) == HARDWARE_TYPE_EUI64)),
+                 error = kThreadError_Parse);
+exit:
+    return error;
+}
+
+ThreadError Dhcp6Server::ProcessServerIdentifier(Message &aMessage, uint16_t aOffset)
+{
+    ThreadError error = kThreadError_None;
+    ServerIdentifier option;
+
+    //aMessage.Read(aOffset, sizeof(option), &option);
+    VerifyOrExit(((aMessage.Read(aOffset, sizeof(option), &option) == sizeof(option)) &&
+                  (option.GetLength() == (sizeof(option) - sizeof(Dhcp6Option))) &&
+                  //VerifyOrExit(((option.GetLength() == (sizeof(option) - sizeof(Dhcp6Option))) &&
+                  (option.GetDuidType() == kDuidLL) &&
+                  (HostSwap16(option.GetDuidLL().mDuidHardwareType) == HARDWARE_TYPE_EUI64) &&
+                  (!memcmp(option.GetDuidLL().mDuidAddrLL, mMac.GetExtAddress(), ADDRESS_LL_LEN))),
+                 error = kThreadError_Parse);
+exit:
+    return error;
+}
+
+ThreadError Dhcp6Server::ProcessIaNa(Message &aMessage, uint16_t aOffset,
+                                     Type aType, ClientIdentifier &aClient, IaNa &aIaNa, IaAddress &aIaAddress)
+{
+    ThreadError error = kThreadError_None;
+    uint16_t optionOffset;
+    uint16_t length;
+
+    VerifyOrExit((aMessage.Read(aOffset, sizeof(aIaNa), &aIaNa) == sizeof(aIaNa)), error = kThreadError_Parse);
+    aOffset += sizeof(aIaNa);
+    length = aIaNa.GetLength();
+    VerifyOrExit((optionOffset = FindOption(aMessage, aOffset, length, Dhcp6::kOptionIaAddress)) > 0, ;);
+    SuccessOrExit(error = ProcessIaAddress(aMessage, optionOffset, aType, aClient, aIaAddress));
+
+exit:
+    return error;
+}
+
+ThreadError Dhcp6Server::ProcessIaAddress(Message &aMessage, uint16_t aOffset, Type aType, ClientIdentifier &aClient,
+                                          IaAddress &aIaAddress)
+{
+    ThreadError error = kThreadError_None;
+    PrefixBuffer *prefix = mPrefixesHead;
+    AddressBuffer *address;
+
+    VerifyOrExit((aMessage.Read(aOffset, sizeof(aIaAddress), &aIaAddress) == sizeof(aIaAddress)),
+                 error = kThreadError_Parse);
+
+    switch (aType)
+    {
+    case kTypeSolicit:
+        while (prefix)
+        {
+            if (otIp6PrefixMatch(&aIaAddress.GetAddress(), &(prefix->GetPrefix().mPrefix)) >= prefix->GetPrefix().mLength)
```

Comment:
I don't think the client gets to specify what prefixes it's interested in receiving IPv6 addresses for.  Nothing in RFC 3315 specifies that behavior.  A simple DHCP server implementation will just assign addresses for all configured prefixes to the client.  A single IA_NA Option can carry multiple IAADDR options.  I think it would be simpler and more efficient to just return all addresses within a single DHCPv6 Reply.